### PR TITLE
[tf.data] Add more precise processing accounting

### PR DESCRIPTION
**Pull request by mkuchnik**
The execution time of a synchronous captured function is currently approximated by the total elapsed time (including queueing time) rather than only measuring execution time. Time in queue is ~10ns per op or more and may increase under load. Mirroring the asynchronous captured function implementation, this commit pushes the accounting into the executor. Processing time now only measures time spent executing.

### CODE REVIEW

**Comment by mkuchnik** [comment]
@jsimsa

File: tensorflow/core/common_runtime/function.cc
Comment (jsimsa) [inline tensorflow/core/common_runtime/function.cc:172]: What is the benefit of passing the callback into the method as opposed to invoking it when the method returns?

File: tensorflow/core/kernels/data/experimental/group_by_reducer_dataset_op.cc
Comment (jsimsa) [inline tensorflow/core/kernels/data/experimental/group_by_reducer_dataset_op.cc:246]: We almost certainly do not want to require callers of `Run` to be aware of the need to wrap the invocation in `RecordStop` and `RecordStart`. Since we are passing `IteratorContext` and `Node` into the method, I can imagine this could be folded into the `Run` method.

**Review by jsimsa — CHANGES_REQUESTED** [review]
Thanks Michael. I have a couple of initial comments. My hope is that the surface that this change touches could be much smaller (ideally restricted to `captured_function.cc`).

File: tensorflow/core/common_runtime/function.cc
Comment (mkuchnik) [reply tensorflow/core/common_runtime/function.cc]: I agree that it is not necessary. I have removed the callback and called the processing directly.

File: tensorflow/core/kernels/data/experimental/group_by_reducer_dataset_op.cc
Comment (mkuchnik) [reply tensorflow/core/kernels/data/experimental/group_by_reducer_dataset_op.cc]: `RecordStop` and `RecordStart` are now called before and after RunSync dispatch.

File: tensorflow/core/kernels/data/captured_function.cc
Comment (jsimsa) [inline tensorflow/core/kernels/data/captured_function.cc:852]: "usage is for function" => "usage for function"

File: tensorflow/core/kernels/data/captured_function.cc
Comment (jsimsa) [inline tensorflow/core/kernels/data/captured_function.cc:816]: Could you apply a similar change to `RunWithBorrowedArgs`? Don't worry about `RunInstantiated` (as that one is only invoked in one place during iterator destruction). Thanks.

File: tensorflow/core/kernels/data/captured_function.cc
Comment (jsimsa) [inline tensorflow/core/kernels/data/captured_function.cc:864]: For consistency with `RunAsync`, this part should be taken out of the `if (collect_usage)` body.

I think it should be:

```
if (node) {
  if (collect_usage) node->record_stop(...);
  TF_RETURN_IF_ERROR(lib_->RunSync(std::move(f_opts), f_handle_, &frame));
  if (ctx->stats_aggregator()) {
    ...
  }
  node->add_processing_time(stats_collector->processing_time());
  if (collect_usage) node->record_start(...);
}
```

**Review by jsimsa — CHANGES_REQUESTED** [review]
Thank you Michael. A couple more minor comments.

File: tensorflow/core/kernels/data/captured_function.cc
Comment (jsimsa) [inline tensorflow/core/kernels/data/captured_function.cc:453]: It would be preferable if we pipe a `Node` through `MakeIteratorFromInputElement` (to correctly account for processing time of transformations such as `flat_map` and `interleave`.

File: tensorflow/core/kernels/data/captured_function.cc
Comment (jsimsa) [inline tensorflow/core/kernels/data/captured_function.cc:909]: Add a `TODO(jsimsa)` for factoring out the common code between `RunAsync`, `Run`, and `RunWithBorrowedArguments` to a function.

File: tensorflow/core/kernels/data/captured_function.h
Comment (jsimsa) [inline tensorflow/core/kernels/data/captured_function.h:224]: Please add an overload for `Run` and `RunWithBorrowedArgs` which does not take `Node` (i.e. matching the signature of the method prior to your change), which calls into the newly introduced API, setting the argument to `nullptr`. That way, you PR will be able to merge cleanly with internal version of TensorFlow without having to updated any internal callers of `Run` and `RunWithBorrowedArgs` synchronously with submitting your OR.

File: tensorflow/core/kernels/data/experimental/io_ops.cc
Comment (jsimsa) [inline tensorflow/core/kernels/data/experimental/io_ops.cc:152]: Does passing simply `nullptr` not work? To improve readability, use the following style:

`function->RunWithBorrowedArgs(ctx, element, &output_tensors, /*node=*/nullptr);`

**Review by jsimsa — CHANGES_REQUESTED** [review]
Thank you Michael. A couple more hopefully final comments.

File: tensorflow/core/kernels/data/captured_function.cc
Comment (mkuchnik) [reply tensorflow/core/kernels/data/captured_function.cc:817]: Done

File: tensorflow/core/kernels/data/captured_function.cc
Comment (mkuchnik) [reply tensorflow/core/kernels/data/captured_function.cc:453]: Sounds good. By the way, the current implementation intentionally ignores functions outside the iterator's GetNext() path (e.g., Initialize code for Snapshot). Those are typically one-off operations and therefore I believe should not fall into processing time but let me know if you think otherwise.

File: tensorflow/core/kernels/data/captured_function.h
Comment (jsimsa) [inline tensorflow/core/kernels/data/captured_function.h:218]: Please do not use default arguments as they make it difficult to extend the API. Instead add the overload as per my suggestion.

File: tensorflow/core/kernels/data/captured_function.cc
Comment (jsimsa) [reply tensorflow/core/kernels/data/captured_function.cc:453]: Using non-null Node only within GetNext makes sense. interleave and flat_map kernels end up calling the method within GetNext.

File: tensorflow/core/kernels/data/captured_function.h
Comment (mkuchnik) [reply tensorflow/core/kernels/data/captured_function.h]: I think Default args are the most clean way to express this here, since this is a non-virtual function and overloading adds the same logical boilerplate.

Comment (mkuchnik) [reply tensorflow/core/kernels/data/captured_function.h]: Ah, just saw the other comment. Will fix.

Comment (jsimsa) [reply tensorflow/core/kernels/data/captured_function.h]: Default arguments are strongly discouraged by Google C++ style guide even for non-virtual functions.

Comment (jsimsa) [reply tensorflow/core/kernels/data/captured_function.h]: https://google.github.io/styleguide/cppguide.html#Default_Arguments

Comment (jsimsa) [reply tensorflow/core/kernels/data/captured_function.h]: For tf.data code we heed this advice and consistency within tf.data code base outweighs the lack of boilerplate default arguments provide.

File: tensorflow/core/kernels/data/captured_function.cc
Comment (mkuchnik) [reply tensorflow/core/kernels/data/captured_function.cc:454]: Does this also require function overloading for internal API purposes?

File: tensorflow/core/kernels/data/parallel_interleave_dataset_op.cc
Comment (mkuchnik) [inline tensorflow/core/kernels/data/parallel_interleave_dataset_op.cc:1002]: Parallel interleave is more complicated with how `RecordStop` and `RecordStart` are used (i.e., in `SaveInternal`).

File: tensorflow/core/kernels/data/captured_function.cc
Comment (jsimsa) [reply tensorflow/core/kernels/data/captured_function.cc:454]: Yes, I was able to find internal calls to this API.

**Review by jsimsa — APPROVED** [review]
LGTM, modulo adding overload for `MakeIteratorFromInputElement`.

**Review by jsimsa — APPROVED** [review]

**Comment by jsimsa** [comment]
gbaned@ could you please let us know what is this PR blocked on? thanks

**Comment by rthadur** [comment]
@jsimsa there was some error while migrating, now this is resolved ,can you please approve this internally.

File: tensorflow/core/kernels/data/captured_function.cc
Comment (rthadur) [inline tensorflow/core/kernels/data/captured_function.cc:451]: Here are the internal errors , can you please check 

```
/tensorflow/core/kernels/data/captured_function.cc:450:50: error: all paths through this function will call itself
    std::unique_ptr<IteratorBase>* out_iterator) {
                                                 ^
1 error generated.
/tensorflow/core/kernels/data/captured_function.cc:450:50: error: all paths through this function will call itself 
    std::unique_ptr<IteratorBase>* out_iterator) {
```

**Review by rthadur — CHANGES_REQUESTED** [review]

File: tensorflow/core/kernels/data/captured_function.cc
Comment (jsimsa) [reply tensorflow/core/kernels/data/captured_function.cc:451]: Michael, you forgot to pass through the `out_iterator` argument.

**Review by jsimsa — APPROVED** [review]

**Pull request merged**
